File: src/app/interests/lifting/exercise/details.js
// The server owns exercise definitions and markup; this adapter opens the
// inspector and moves existing rows between selected values and the picker.
const dialog = document.querySelector('[data-exercise-details-dialog]');
if (dialog) initDetails(dialog);

function initDetails(dialog) {
  const body = dialog.querySelector('[data-exercise-details-body]');
  const close = dialog.querySelector('[data-exercise-details-close]');
  let request, generation = 0, trigger;
  const initiallyOpen = dialog.open;
  if (initiallyOpen) dialog.removeAttribute('open');

  function selectionUrl(name, opened = true) {
    const url = new URL(location.href);
    url.pathname = '/fitness/exercises';
    url.searchParams.set('exercise', name);
    url.searchParams.delete('notice');
    if (opened) url.searchParams.set('details', '1'); else url.searchParams.delete('details');
    return url;
  }

  function install(content) {
    body.replaceChildren(document.importNode(content, true));
    enhanceEditor(body);
    const name = content.dataset.exerciseName;
    if (name) {
      document.dispatchEvent(new CustomEvent('exercise-details-select', {detail:{name}}));
      history.replaceState(null, '', selectionUrl(name));
      close.href = selectionUrl(name, false);
    }
    body.querySelector('[data-details-heading]')?.focus({preventScroll:true});
  }

  async function open(name, source) {
    trigger = source || trigger;
    request?.abort();
    const current = ++generation;
    request = new AbortController();
    const signal = request.signal;
    document.dispatchEvent(new CustomEvent('exercise-details-select', {detail:{name}}));
    history.replaceState(null, '', selectionUrl(name));
    close.href = selectionUrl(name, false);
    body.replaceChildren(document.querySelector('[data-exercise-details-loading]')?.content.cloneNode(true) || document.createTextNode('Loading exercise…'));
    dialog.setAttribute('aria-busy', 'true');
    if (!dialog.open) dialog.showModal();
    try {
      const url = new URL('/fitness/exercises/details', location.href);
      url.searchParams.set('exercise', name);
      const response = await fetch(url, {signal, credentials:'same-origin', headers:{Accept:'text/html'}});
      if (signal.aborted || current !== generation) return;
      const html = new DOMParser().parseFromString(await response.text(), 'text/html');
      if (signal.aborted || current !== generation) return;
      const content = html.querySelector('[data-exercise-details-content]');
      if (!response.ok || !content) throw new Error('Exercise details could not load. Close and try again.');
      install(content);
    } catch (error) {
      if (!signal.aborted && current === generation) body.textContent = error.message;
    } finally {
      if (current === generation) dialog.removeAttribute('aria-busy');
    }
  }

  document.addEventListener('click', (event) => {
    const link = event.target.closest('[data-exercise-details-link]');
    if (!link || event.button !== 0 || event.ctrlKey || event.metaKey || event.shiftKey || event.altKey) return;
    const name = new URL(link.href).searchParams.get('exercise');
    if (!name) return;
    event.preventDefault();
    void open(name, link);
  });
  close.addEventListener('click', (event) => { event.preventDefault(); dialog.close(); });
  dialog.addEventListener('close', () => {
    request?.abort(); generation++;
    dialog.removeAttribute('aria-busy');
    const url = new URL(location.href);
    if (url.pathname !== '/fitness/exercises') { location.assign(close.href); return; }
    url.searchParams.delete('details'); url.searchParams.delete('notice');
    history.replaceState(null, '', url);
    trigger?.focus({preventScroll:true});
  });
  window.addEventListener('popstate', () => {
    const url = new URL(location.href);
    if (url.searchParams.get('details') === '1' && url.searchParams.get('exercise')) void open(url.searchParams.get('exercise'));
    else if (dialog.open) dialog.close();
  });
  dialog.addEventListener('submit', (event) => {
    const form = event.target.closest('[data-exercise-definition-form], [data-exercise-identity-form]');
    if (!form) return;
    event.preventDefault();
    if (form.dataset.busy !== 'true') void save(form);
  });

  async function save(form) {
    const identity = form.hasAttribute('data-exercise-identity-form');
    const status = form.querySelector('[data-editor-status]');
    const payload = new URLSearchParams(new FormData(form));
    const controls = [...form.querySelectorAll('button')];
    form.dataset.busy = 'true'; controls.forEach((button) => { button.disabled = true; });
    status.textContent = identity ? 'Checking name and aliases…' : 'Saving changes…';
    try {
      const response = await fetch(form.action, {method:'POST',credentials:'same-origin',body:payload,headers:{Accept:identity ? 'text/html' : 'application/json','X-Exercise-Dialog':'1'}});
      if (!response.ok) throw new Error(await response.text() || 'Changes could not be saved. Try again.');
      if (!identity) {
        const saved = await response.json();
        if (typeof saved.name !== 'string') throw new Error('The save response was incomplete. Retry to check the saved exercise.');
        const url = selectionUrl(saved.name);
        url.searchParams.set('notice', 'saved');
        location.assign(url);
        return;
      }
      if (response.redirected) {
        const url = new URL(response.url);
        if (url.pathname !== '/fitness/exercises' || url.searchParams.get('notice') !== 'identity-saved') {
          throw new Error(url.searchParams.get('notice') === 'identity-stale' ? 'The archive changed. Review the name and aliases again.' : 'Changes were not saved. Check your sign-in and try again.');
        }
        const destination = selectionUrl(url.searchParams.get('exercise'));
        destination.searchParams.set('notice', 'identity-saved');
        location.assign(destination);
        return;
      }
      const content = new DOMParser().parseFromString(await response.text(), 'text/html').querySelector('[data-exercise-details-content]');
      if (!content) throw new Error('The review could not load. Try again.');
      install(content);
    } catch (error) {
      status.textContent = error.message || 'Changes could not be saved. Try again.';
    } finally {
      form.dataset.busy = 'false'; controls.forEach((button) => { button.disabled = false; });
    }
  }

  if (initiallyOpen) { dialog.showModal(); enhanceEditor(body); body.querySelector('[data-details-heading]')?.focus({preventScroll:true}); }
}

function enhanceEditor(body) {
  body.dataset.editorEnhanced = 'true';
  const form = body.querySelector('[data-exercise-definition-form]');
  if (!form) return;
  function updateDiagram() {
    const weights = new Map([...form.querySelectorAll('[data-editor-group="muscle"] [data-editor-selected] [data-editor-item]')].map((row) => [row.dataset.editorItem, Number(row.querySelector('input')?.value || row.querySelector('strong')?.textContent || 0)]));
    for (const path of body.querySelectorAll('.exercise-details__body-map [data-muscle]')) {
      const weight = weights.get(path.dataset.muscle) || 0;
      path.dataset.weight = weight >= 75 ? 'primary' : weight > 0 ? 'secondary' : 'none';
    }
  }
  for (const group of form.querySelectorAll('[data-editor-group]')) {
    const selected = group.querySelector('[data-editor-selected]');
    const available = group.querySelector('[data-editor-available]');
    if (!available) continue;
    const search = group.querySelector('[data-editor-search]');
    search.hidden = false;
    function refresh() {
      const terms = search.value.toLowerCase().trim().split(/\s+/).filter(Boolean);
      for (const row of group.querySelectorAll('[data-editor-item]')) {
        const active = row.parentElement === selected;
        const input = row.querySelector('[data-editor-value]');
        input.disabled = !active;
        if (input.type === 'checkbox') input.checked = active;
        row.querySelector('[data-editor-add]').hidden = active;
        row.querySelector('[data-editor-remove]').hidden = !active;
        row.hidden = !active && !terms.every((term) => row.dataset.editorTerms.toLowerCase().includes(term));
      }
      group.querySelector('[data-editor-no-matches]').hidden = [...available.children].some((row) => !row.hidden);
      updateDiagram();
    }
    group.addEventListener('click', (event) => {
      const button = event.target.closest('[data-editor-add], [data-editor-remove]');
      if (!button) return;
      const row = button.closest('[data-editor-item]');
      const adding = button.hasAttribute('data-editor-add');
      (adding ? selected : available).append(row);
      const input = row.querySelector('[data-editor-value]');
      if (adding && input.type === 'number' && Number(input.value) <= 0) input.value = '100';
      refresh();
      if (adding) {
        if (input.type === 'number') { input.focus(); input.select(); } else row.querySelector('[data-editor-remove]').focus();
      } else group.querySelector('summary').focus();
    });
    search.addEventListener('input', refresh);
    search.addEventListener('keydown', (event) => {
      if (event.key !== 'Enter') return;
      event.preventDefault();
      [...available.children].find((row) => !row.hidden)?.querySelector('[data-editor-add]').click();
    });
    group.querySelector('details').addEventListener('toggle', (event) => { if (event.target.open) search.focus(); });
    refresh();
  }
  form.addEventListener('input', updateDiagram);
  updateDiagram();
}

File: src/app/interests/lifting/exercise/details.rs
//! One server-rendered exercise inspector for both catalog views and dialog fetches.
use topcoat::{
    Result,
    asset::{Asset, asset},
    context::{Cx, app_context},
    router::{
        HeaderValue, StatusCode, header, query_params,
        response::{IntoResponse, Response},
        route,
    },
    view::{component, view},
};

use super::super::{
    archive::{exercise_definition::Definition, snapshot::ExerciseProfile, store::FitnessStore},
    filters::{EQUIPMENT, MOVEMENT_DETAILS, MOVEMENTS},
    muscle_taxonomy, muscles,
};
use crate::{app::login::viewer, content::access::is_admin};

pub(in super::super) const DETAILS_JS: Asset = asset!("./details.js");

#[query_params(error = redirect("?"))]
struct DetailsQuery {
    view: Option<String>,
    exercise: Option<String>,
    details: Option<String>,
    notice: Option<String>,
}

#[component]
pub(in super::super) async fn host(cx: &Cx) -> Result {
    let query = query_params::<DetailsQuery>(cx)?;
    let name = query
        .exercise
        .as_deref()
        .filter(|name| super::plausible_exercise_name(name));
    let opened = query.details.as_deref() == Some("1") && name.is_some();
    let close_url = if query.view.as_deref() == Some("list") {
        "/fitness/exercises?view=list".to_string()
    } else {
        name.map_or_else(
            || super::super::exercise_space::PATH.to_string(),
            super::page_url,
        )
    };
    view! {
        <dialog class="exercise-details" data-exercise-details-dialog="" aria-label="Exercise details" open=(opened)>
            <a class="exercise-details__close" href=(close_url.as_str()) data-exercise-details-close="" aria-label="Close exercise details">"×"</a>
            <div data-exercise-details-body="">
                if opened { content(name: name.unwrap(), notice: query.notice.as_deref()) }
            </div>
        </dialog>
        <template data-exercise-details-loading=""><p class="exercise-details__message" role="status">"Loading exercise…"</p></template>
        <script type="module" src=(DETAILS_JS)></script>
    }
}

#[route(GET "/fitness/exercises/details")]
async fn details_endpoint(cx: &Cx) -> Result<Response> {
    let query = query_params::<DetailsQuery>(cx)?;
    let Some(name) = query
        .exercise
        .as_deref()
        .filter(|name| super::plausible_exercise_name(name))
    else {
        return Ok(super::plain(StatusCode::BAD_REQUEST, "Choose an exercise."));
    };
    view! {
        ((header::CACHE_CONTROL, HeaderValue::from_static("no-store")))
        content(name: name, notice: query.notice.as_deref())
    }?
    .into_response(cx)
}

#[component]
async fn content(cx: &Cx, name: &str, notice: Option<&str>) -> Result {
    let Ok(snapshot) = app_context::<FitnessStore>(cx).snapshot().await else {
        return view! { (StatusCode::SERVICE_UNAVAILABLE) <p class="exercise-details__message">"Exercise details could not load. Try again."</p> };
    };
    let Some(name) = snapshot.canonical_exercise_name(name) else {
        return view! { (StatusCode::NOT_FOUND) <p class="exercise-details__message">"That exercise was not found."</p> };
    };
    let definition = Definition::from_snapshot(&snapshot, &name);
    let history = snapshot.exercise_profile(&name);
    let aliases = snapshot.exercise_aliases(&name);
    let can_edit = viewer(cx).is_some_and(|current| is_admin(&current.email));
    let provenance = if can_edit {
        super::db_sources(cx, &name)
            .await
            .ok()
            .and_then(|sources| super::provenance_line(&sources))
    } else {
        None
    };
    view! {
        inspector(definition: &definition, aliases: &aliases, can_edit: can_edit, provenance: provenance.as_deref(), notice: notice, history: history.as_ref())
    }
}

#[component]
async fn inspector(
    definition: &Definition,
    aliases: &[String],
    can_edit: bool,
    provenance: Option<&str>,
    notice: Option<&str>,
    #[default(None)] history: Option<&ExerciseProfile>,
) -> Result {
    let mut involvement = muscles::MuscleInvolvement::default();
    for (id, _) in muscle_taxonomy::muscles() {
        match definition.weights.get(id).copied().unwrap_or(0) {
            75..=100 => involvement.primary.push(id),
            1..=74 => involvement.secondary.push(id),
            _ => {}
        }
    }
    let notice = if can_edit {
        notice.and_then(|code| match code {
            "saved" => Some("Changes saved."),
            "identity-saved" => Some("Name and aliases saved."),
            "identity-stale" => Some("The archive changed. Review the name and aliases again."),
            "invalid" => Some("That did not validate. Nothing changed."),
            "unavailable" => Some("The exercise could not be saved. Try again."),
            _ => None,
        })
    } else {
        None
    };
    view! {
        <section data-exercise-details-content="" data-exercise-name=(definition.name.as_str())>
            <header class="exercise-details__header">
                <div><h2 tabindex="-1" data-details-heading="">(definition.name.as_str())</h2>
                    if let Some(history) = history {
                        <p title=(format!("{} through {}", history.first_date, history.last_date))>
                            (format!("{} {} across {} {}", history.set_count, super::super::format::plural(history.set_count, "set", "sets"), history.workout_count, super::super::format::plural(history.workout_count, "workout", "workouts")))
                        </p>
                    } else { <p>"No sets logged yet"</p> }
                    if !aliases.is_empty() { <p>(format!("Also known as {}", aliases.join(", ")))</p> }
                    if !can_edit { <p>"Muscle weights and exercise setup"</p> }
                </div>
                <div class="exercise-details__body-map" aria-label="Muscles worked">
                    muscles::muscle_figure(paths: muscles::FRONT_PATHS, caption: "front", involvement: &involvement, compact: true)
                    muscles::muscle_figure(paths: muscles::BACK_PATHS, caption: "back", involvement: &involvement, compact: true)
                </div>
            </header>
            if let Some(notice) = notice { <p class="exercise-details__notice" role="status">(notice)</p> }
            <form action=(format!("{}/definition", super::write_url(&definition.name))) method="post" data-exercise-definition-form="">
                if can_edit {
                    <input type="hidden" name="name" value=(definition.name.as_str())>
                    <input type="hidden" name="editing" value=(definition.name.as_str())>
                }
                <section class="exercise-editor-group" data-editor-group="muscle">
                    <h3>"Muscles"</h3>
                    <p class="exercise-details__hint">"100 = full credit · 50 = half. Weights don’t need to total 100."</p>
                    <div class="exercise-editor-muscles" data-editor-selected="">
                        for (id, label) in muscle_taxonomy::muscles() {
                            if let Some(ratio) = definition.weights.get(id).filter(|ratio| **ratio > 0) {
                                muscle_row(id: id, label: label, ratio: *ratio, can_edit: can_edit)
                            }
                        }
                    </div>
                    <p class="exercise-editor-empty">"No muscles added."</p>
                    if can_edit {
                        <details class="exercise-editor-add" data-editor-picker="">
                            <summary>"+ Add muscles"</summary>
                            <input type="search" aria-label="Search muscles" placeholder="Search muscles or body regions" data-editor-search="" hidden="">
                            <div class="exercise-editor-options" data-editor-available="">
                                for (id, label) in muscle_taxonomy::muscles() {
                                    if !definition.weights.contains_key(id) { muscle_row(id: id, label: label, ratio: 0, can_edit: true) }
                                }
                            </div>
                            <p class="exercise-details__hint" data-editor-no-matches="" hidden="">"No more matching muscles."</p>
                        </details>
                    }
                </section>
                tag_group(kind: "movement", heading: "Movement patterns", options: MOVEMENTS.iter().chain(MOVEMENT_DETAILS).copied().collect(), selected: &definition.movements, can_edit: can_edit)
                tag_group(kind: "equipment", heading: "Equipment", options: EQUIPMENT.to_vec(), selected: &definition.equipment, can_edit: can_edit)
                if can_edit {
                    <footer class="exercise-details__save"><p role="status" data-editor-status=""></p><button class="entry-button entry-button--primary" type="submit">"Save changes"</button></footer>
                }
            </form>
            if can_edit {
                <details class="exercise-details__identity"><summary>"Name & aliases"</summary>
                    super::identity_form(name: definition.name.as_str(), aliases: aliases)
                    if let Some(provenance) = provenance { <p class="exercise-details__hint">(provenance)</p> }
                </details>
            }
        </section>
    }
}

#[component]
async fn muscle_row(id: &str, label: &str, ratio: u32, can_edit: bool) -> Result {
    let region = muscle_taxonomy::coarse_tag_for(id).unwrap_or("");
    view! {
        <div class="exercise-editor-muscle" data-editor-item=(id) data-editor-terms=(format!("{label} {id} {region}"))>
            <label for=(format!("detail-ratio-{id}"))>(label)</label>
            if can_edit {
                <input id=(format!("detail-ratio-{id}")) name=(format!("ratio_{id}")) type="number" min="0" max="100" step="1" inputmode="numeric" value=(ratio) data-editor-value="" aria-label=(format!("{label} weight"))>
                <button type="button" data-editor-add="" aria-label=(format!("Add {label}")) hidden="">"+"</button>
                <button type="button" data-editor-remove="" aria-label=(format!("Remove {label}")) hidden="">"×"</button>
            } else { <strong>(ratio)</strong> }
        </div>
    }
}

#[component]
async fn tag_group(
    kind: &str,
    heading: &str,
    options: Vec<(&str, &str)>,
    selected: &[String],
    can_edit: bool,
) -> Result {
    view! {
        <section class="exercise-editor-group" data-editor-group=(kind)>
            <h3>(heading)</h3>
            <div class="exercise-editor-tags" data-editor-selected="">
                for (id, label) in &options {
                    if selected.iter().any(|value| value == id) { tag_row(kind: kind, id: id, label: label, selected: true, can_edit: can_edit) }
                }
            </div>
            <p class="exercise-editor-empty">"None added."</p>
            if can_edit {
                <details class="exercise-editor-add" data-editor-picker="">
                    <summary>(format!("+ Add {}", heading.to_lowercase()))</summary>
                    <input type="search" aria-label=(format!("Search {}", heading.to_lowercase())) placeholder=(format!("Search {}", heading.to_lowercase())) data-editor-search="" hidden="">
                    <div class="exercise-editor-options" data-editor-available="">
                        for (id, label) in &options {
                            if !selected.iter().any(|value| value == id) { tag_row(kind: kind, id: id, label: label, selected: false, can_edit: true) }
                        }
                    </div>
                    <p class="exercise-details__hint" data-editor-no-matches="" hidden="">"No more matching options."</p>
                </details>
            }
        </section>
    }
}

#[component]
async fn tag_row(kind: &str, id: &str, label: &str, selected: bool, can_edit: bool) -> Result {
    view! {
        <div class="exercise-editor-tag" data-editor-item=(id) data-editor-terms=(format!("{label} {id}"))>
            <label>
                if can_edit { <input type="checkbox" name=(kind) value=(id) checked=(selected) data-editor-value=""> }
                <span>(label)</span>
            </label>
            if can_edit {
                <button type="button" data-editor-add="" aria-label=(format!("Add {label}")) hidden="">"+"</button>
                <button type="button" data-editor-remove="" aria-label=(format!("Remove {label}")) hidden="">"×"</button>
            }
        </div>
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn inspector_exposes_only_existing_values_until_expanded_and_never_public_writes() {
        let definition = Definition {
            name: "Press & pull".into(),
            weights: [("biceps".into(), 65)].into(),
            movements: vec!["horizontal-pull".into()],
            ..Definition::default()
        };
        let cx = Cx::default();
        let __cx = &cx;
        let result: Result = view! { inspector(definition: &definition, aliases: &[], can_edit: false, provenance: None, notice: None) };
        let public = result.unwrap().render(__cx);
        assert!(public.contains("Press &amp; pull"));
        assert!(public.contains("<strong>65</strong>"));
        assert!(!public.contains("name=\"ratio_"));
        assert!(!public.contains("Save changes"));
        assert!(!public.contains("data-editor-picker"));
        let result: Result = view! { inspector(definition: &definition, aliases: &[], can_edit: true, provenance: None, notice: None) };
        let owner = result.unwrap().render(__cx);
        assert!(owner.contains("data-editor-picker"));
        assert_eq!(owner.matches("name=\"ratio_biceps\"").count(), 1);
        assert!(owner.contains("value=\"65\""));
        assert!(owner.contains("data-exercise-identity-form"));
    }
}
